File: frontend/lib/features/workflows/domain/workflow.dart
class WorkflowUserSummary {
  final String id;
  final String email;
  final String? firstName;
  final String? lastName;

  const WorkflowUserSummary({
    required this.id,
    required this.email,
    this.firstName,
    this.lastName,
  });

  String get displayName {
    final parts = [
      firstName,
      lastName,
    ].whereType<String>().where((p) => p.isNotEmpty);
    if (parts.isNotEmpty) return parts.join(' ');
    return email;
  }

  factory WorkflowUserSummary.fromJson(Map<String, dynamic> json) {
    return WorkflowUserSummary(
      id: json['id'] as String,
      email: json['email'] as String? ?? '',
      firstName: json['first_name'] as String?,
      lastName: json['last_name'] as String?,
    );
  }
}

class WorkflowHistoryEntry {
  final String id;
  final String action;
  final String? comment;
  final String? fromState;
  final String? toState;
  final int escalationLevel;
  final DateTime createdAt;
  final WorkflowUserSummary? actor;
  final WorkflowUserSummary? assignee;

  const WorkflowHistoryEntry({
    required this.id,
    required this.action,
    this.comment,
    this.fromState,
    this.toState,
    required this.escalationLevel,
    required this.createdAt,
    this.actor,
    this.assignee,
  });

  factory WorkflowHistoryEntry.fromJson(Map<String, dynamic> json) {
    return WorkflowHistoryEntry(
      id: json['id'] as String,
      action: json['action'] as String? ?? '',
      comment: json['comment'] as String?,
      fromState: json['from_state'] as String?,
      toState: json['to_state'] as String?,
      escalationLevel: json['escalation_level'] as int? ?? 0,
      createdAt: DateTime.parse(json['created_at'] as String),
      actor: json['actor'] != null
          ? WorkflowUserSummary.fromJson(json['actor'] as Map<String, dynamic>)
          : null,
      assignee: json['assignee'] != null
          ? WorkflowUserSummary.fromJson(
              json['assignee'] as Map<String, dynamic>,
            )
          : null,
    );
  }
}

class WorkflowComment {
  final String id;
  final String body;
  final DateTime createdAt;
  final WorkflowUserSummary? author;

  const WorkflowComment({
    required this.id,
    required this.body,
    required this.createdAt,
    this.author,
  });

  factory WorkflowComment.fromJson(Map<String, dynamic> json) {
    return WorkflowComment(
      id: json['id'] as String,
      body: json['body'] as String? ?? '',
      createdAt: DateTime.parse(json['created_at'] as String),
      author: json['author'] != null
          ? WorkflowUserSummary.fromJson(json['author'] as Map<String, dynamic>)
          : null,
    );
  }
}

class WorkflowRequest {
  final String id;
  final String requestNumber;
  final String title;
  final String? description;
  final String requestType;
  final String state;
  final String requesterId;
  final WorkflowUserSummary? requester;
  final String? assigneeId;
  final WorkflowUserSummary? assignee;
  final String? entityType;
  final String? entityId;
  final Map<String, dynamic>? payload;
  final int escalationLevel;
  final DateTime? pendingSince;
  final DateTime? submittedAt;
  final DateTime? resolvedAt;
  final DateTime createdAt;
  final DateTime updatedAt;
  final List<WorkflowHistoryEntry> history;
  final List<WorkflowComment> comments;

  const WorkflowRequest({
    required this.id,
    required this.requestNumber,
    required this.title,
    this.description,
    required this.requestType,
    required this.state,
    required this.requesterId,
    this.requester,
    this.assigneeId,
    this.assignee,
    this.entityType,
    this.entityId,
    this.payload,
    required this.escalationLevel,
    this.pendingSince,
    this.submittedAt,
    this.resolvedAt,
    required this.createdAt,
    required this.updatedAt,
    this.history = const [],
    this.comments = const [],
  });

  bool get isDraft => state == 'draft';
  bool get isPending => state == 'pending';
  bool get isTerminal => state == 'approved' || state == 'rejected';

  factory WorkflowRequest.fromJson(Map<String, dynamic> json) {
    return WorkflowRequest(
      id: json['id'] as String,
      requestNumber: json['request_number'] as String? ?? '',
      title: json['title'] as String? ?? '',
      description: json['description'] as String?,
      requestType: json['request_type'] as String? ?? 'general',
      state: json['state'] as String? ?? 'draft',
      requesterId: json['requester_id'] as String,
      requester: json['requester'] != null
          ? WorkflowUserSummary.fromJson(
              json['requester'] as Map<String, dynamic>,
            )
          : null,
      assigneeId: json['assignee_id'] as String?,
      assignee: json['assignee'] != null
          ? WorkflowUserSummary.fromJson(
              json['assignee'] as Map<String, dynamic>,
            )
          : null,
      entityType: json['entity_type'] as String?,
      entityId: json['entity_id'] as String?,
      payload: json['payload'] as Map<String, dynamic>?,
      escalationLevel: json['escalation_level'] as int? ?? 0,
      pendingSince: json['pending_since'] != null
          ? DateTime.parse(json['pending_since'] as String)
          : null,
      submittedAt: json['submitted_at'] != null
          ? DateTime.parse(json['submitted_at'] as String)
          : null,
      resolvedAt: json['resolved_at'] != null
          ? DateTime.parse(json['resolved_at'] as String)
          : null,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      history:
          (json['history'] as List<dynamic>?)
              ?.map(
                (e) => WorkflowHistoryEntry.fromJson(e as Map<String, dynamic>),
              )
              .toList() ??
          const [],
      comments:
          (json['comments'] as List<dynamic>?)
              ?.map((e) => WorkflowComment.fromJson(e as Map<String, dynamic>))
              .toList() ??
          const [],
    );
  }

  Map<String, dynamic> toCreateJson() => {
    'title': title,
    if (description != null) 'description': description,
    'request_type': requestType,
    if (entityType != null) 'entity_type': entityType,
    if (entityId != null) 'entity_id': entityId,
    if (payload != null) 'payload': payload,
    if (assigneeId != null) 'assignee_id': assigneeId,
  };

  Map<String, dynamic> toUpdateJson() => {
    'title': title,
    if (description != null) 'description': description,
    'request_type': requestType,
    if (entityType != null) 'entity_type': entityType,
    if (entityId != null) 'entity_id': entityId,
    if (payload != null) 'payload': payload,
  };
}

class PaginatedWorkflowRequests {
  final List<WorkflowRequest> items;
  final int total;
  final int skip;
  final int limit;

  const PaginatedWorkflowRequests({
    required this.items,
    required this.total,
    required this.skip,
    required this.limit,
  });

  factory PaginatedWorkflowRequests.fromJson(Map<String, dynamic> json) {
    final items =
        (json['items'] as List<dynamic>?)
            ?.map((e) => WorkflowRequest.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [];
    return PaginatedWorkflowRequests(
      items: items,
      total: json['total'] as int? ?? items.length,
      skip: json['skip'] as int? ?? 0,
      limit: json['limit'] as int? ?? items.length,
    );
  }
}

String workflowStateLabel(String state) {
  switch (state) {
    case 'draft':
      return 'Draft';
    case 'pending':
      return 'Pending';
    case 'approved':
      return 'Approved';
    case 'rejected':
      return 'Rejected';
    default:
      return state;
  }
}

String workflowTypeLabel(String type) {
  switch (type) {
    case 'general':
      return 'General';
    case 'transaction':
      return 'Transaction';
    case 'inventory':
      return 'Inventory';
    case 'customer':
      return 'Customer';
    default:
      return type;
  }
}

String workflowActionLabel(String action) {
  switch (action) {
    case 'created':
      return 'Created';
    case 'submitted':
      return 'Submitted';
    case 'assigned':
      return 'Assigned';
    case 'approved':
      return 'Approved';
    case 'rejected':
      return 'Rejected';
    case 'escalated':
      return 'Escalated';
    default:
      return action;
  }
}

File: backend/app/schemas/workflow.py
import uuid
from datetime import datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field

WorkflowState = Literal["draft", "pending", "approved", "rejected"]
WorkflowRequestType = Literal["general", "transaction", "inventory", "customer"]
WorkflowHistoryAction = Literal[
    "created", "submitted", "assigned", "approved", "rejected", "escalated"
]
WorkflowSortField = Literal[
    "request_number",
    "title",
    "state",
    "request_type",
    "created_at",
    "submitted_at",
    "pending_since",
]
SortOrder = Literal["asc", "desc"]


class WorkflowUserSummary(BaseModel):
    id: uuid.UUID
    email: str
    first_name: Optional[str] = None
    last_name: Optional[str] = None

    model_config = {"from_attributes": True}

    @property
    def display_name(self) -> str:
        parts = [p for p in (self.first_name, self.last_name) if p]
        return " ".join(parts) if parts else self.email


class WorkflowHistoryEntry(BaseModel):
    id: uuid.UUID
    actor_id: Optional[uuid.UUID] = None
    actor: Optional[WorkflowUserSummary] = None
    action: str
    comment: Optional[str] = None
    from_state: Optional[str] = None
    to_state: Optional[str] = None
    assignee_id: Optional[uuid.UUID] = None
    assignee: Optional[WorkflowUserSummary] = None
    escalation_level: int
    created_at: datetime

    model_config = {"from_attributes": True}


class WorkflowCommentResponse(BaseModel):
    id: uuid.UUID
    author_id: uuid.UUID
    author: Optional[WorkflowUserSummary] = None
    body: str
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class WorkflowRequestCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=5000)
    request_type: WorkflowRequestType = "general"
    entity_type: Optional[str] = Field(None, max_length=50)
    entity_id: Optional[uuid.UUID] = None
    payload: Optional[dict[str, Any]] = None
    assignee_id: Optional[uuid.UUID] = None


class WorkflowRequestUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=5000)
    request_type: Optional[WorkflowRequestType] = None
    entity_type: Optional[str] = Field(None, max_length=50)
    entity_id: Optional[uuid.UUID] = None
    payload: Optional[dict[str, Any]] = None


class WorkflowSubmitRequest(BaseModel):
    assignee_id: Optional[uuid.UUID] = None
    comment: Optional[str] = Field(None, max_length=2000)


class WorkflowAssignRequest(BaseModel):
    assignee_id: uuid.UUID
    comment: Optional[str] = Field(None, max_length=2000)


class WorkflowDecisionRequest(BaseModel):
    comment: Optional[str] = Field(None, max_length=2000)


class WorkflowCommentCreate(BaseModel):
    body: str = Field(..., min_length=1, max_length=5000)


class WorkflowRequestSummary(BaseModel):
    id: uuid.UUID
    request_number: str
    title: str
    description: Optional[str] = None
    request_type: str
    state: str
    requester_id: uuid.UUID
    requester: Optional[WorkflowUserSummary] = None
    assignee_id: Optional[uuid.UUID] = None
    assignee: Optional[WorkflowUserSummary] = None
    entity_type: Optional[str] = None
    entity_id: Optional[uuid.UUID] = None
    escalation_level: int
    pending_since: Optional[datetime] = None
    submitted_at: Optional[datetime] = None
    resolved_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class WorkflowRequestDetailResponse(WorkflowRequestSummary):
    payload: Optional[dict[str, Any]] = None
    history: list[WorkflowHistoryEntry] = []
    comments: list[WorkflowCommentResponse] = []


class WorkflowEscalationRuleCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    request_type: str = Field(default="*", max_length=50)
    hours_until_escalation: int = Field(..., gt=0, le=8760)
    target_permission: str = Field(..., min_length=1, max_length=100)
    escalation_level: int = Field(default=0, ge=0)
    is_active: bool = True


class WorkflowEscalationRuleUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    request_type: Optional[str] = Field(None, max_length=50)
    hours_until_escalation: Optional[int] = Field(None, gt=0, le=8760)
    target_permission: Optional[str] = Field(None, min_length=1, max_length=100)
    escalation_level: Optional[int] = Field(None, ge=0)
    is_active: Optional[bool] = None


class WorkflowEscalationRuleResponse(BaseModel):
    id: uuid.UUID
    name: str
    request_type: str
    hours_until_escalation: int
    target_permission: str
    escalation_level: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class WorkflowEscalationResult(BaseModel):
    escalated_count: int
    request_ids: list[uuid.UUID]
